File: objectModel/TypeScript/Persistence/CdmFolder/ArgumentPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmArgumentDefinition,
    CdmCorpusContext,
    cdmObjectType,
    copyOptions,
    resolveOptions,
    StringUtils
} from '../../internal';
import {
    Argument,
    CdmJsonType
} from './types';
import * as utils from './utils';

export class ArgumentPersistence {
    public static fromData(ctx: CdmCorpusContext, object: string | Argument): CdmArgumentDefinition {
        const argument: CdmArgumentDefinition = ctx.corpus.MakeObject(cdmObjectType.argumentDef);
        
        // must use `object !== undefined` here, since when object is null, null !== undefined => false but null !== undefined => true
        if (typeof object === 'object' && object !== undefined && object.value !== undefined) {
            argument.value = utils.createConstant(ctx, object.value as CdmJsonType);
            if (object.name) {
                argument.name = object.name;
            }
            if (object.explanation) {
                argument.explanation = object.explanation;
            }
        } else {
            // not a structured argument, just a thing. try it
            argument.value = utils.createConstant(ctx, object);
        }

        return argument;
    }

    public static toData(instance: CdmArgumentDefinition, resOpt: resolveOptions, options: copyOptions): CdmJsonType {
        let value: CdmJsonType;
        if (instance.value !== undefined) {
            if (typeof(instance.value) === 'object' && 'copyData' in instance.value
                && typeof(instance.value.copyData) === 'function') {
                value = instance.value.copyData(resOpt, options);
            } else {
                value = instance.value;
            }
        }
        // skip the argument if just a value
        if (StringUtils.isBlankByCdmStandard(instance.name)) {
            return value;
        }

        return { explanation: instance.explanation, name: instance.name, value: value };
    }
}

File: objectModel/Python/cdm/persistence/cdmfolder/argument_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Union, TYPE_CHECKING

from cdm.enums import CdmObjectType
from cdm.persistence import PersistenceLayer

from . import utils
from .types import Argument, CdmJsonType

if TYPE_CHECKING:
    from cdm.objectmodel import CdmArgumentDefinition, CdmCorpusContext
    from cdm.utilities import CopyOptions, ResolveOptions


class ArgumentPersistence:
    @staticmethod
    def from_data(ctx: 'CdmCorpusContext', data: Union[str, Argument]) -> 'CdmArgumentDefinition':
        from cdm.utilities import JObject

        argument = ctx.corpus.make_object(CdmObjectType.ARGUMENT_DEF)  # type: CdmArgumentDefinition

        if isinstance(data, JObject) and data.get('value') is not None:
            argument.value = utils.create_constant(ctx, data.value)

            if data.get('name'):
                argument.name = data.name

            if data.get('explanation'):
                argument.explanation = data.explanation
        else:
            # Not a structured argument, just a thing. Try it
            argument.value = utils.create_constant(ctx, data)

        return argument

    @staticmethod
    def to_data(instance: 'CdmArgumentDefinition', res_opt: 'ResolveOptions', options: 'CopyOptions') -> CdmJsonType:
        from cdm.objectmodel import CdmObject

        value = None

        if instance.value is not None:
            if isinstance(instance.value, CdmObject):
                value = PersistenceLayer.to_data(instance.value, res_opt, options, PersistenceLayer.CDM_FOLDER)
            else:
                value = instance.value

        # Skip the argument if just a value
        if not instance.name:
            return value

        arg = Argument()
        arg.explanation = instance.explanation
        arg.name = instance.name
        arg.value = value
        return arg
